gtsam/hybrid/DCFactor.h

```diff
@@ -99,10 +111,24 @@ class DCFactor : public Factor {
    * @param discreteVals - Likewise, assignment to the discrete variables in
    * `discreteKeys__`.
    */
-  virtual boost::shared_ptr<GaussianFactor> linearize(
+  virtual GaussianFactor::shared_ptr linearize(
       const Values& continuousVals,
       const DiscreteValues& discreteVals) const = 0;
 
+  /**
+   * @brief Linearize all the continuous factors only with respect to the
+   * continuous variables (as given in `keys_`).
+   *
+   * This `linearize` is different from the `linearize(continuous, discrete)` in
+   * that it assumes no assignment of discrete keys and linearizes all
```

Comment:
Comment on the return type. In fact, should it also be a GaussianMixture?

Reply:
DCFactor is the right type since it is the polymorphic base type.

Reply:
I don't understand this, can it be anything else, ever? Make the type as specific as it can be.

Reply:
So we can use DCGaussianMixtureFactor as the return type, but we've been using DCFactor since it is the common base class for all the DC factors. If we update this to use DCGaussianMixtureFactor, then it becomes a little messy since the forward declaration doesn't exactly work and there will be circular includes if we import DCGaussianMixtureFactor in DCFactor.

Reply:
Did you actually investigate or are you shooting from the hip? I think it's a valid review comment.
Why can you not just say 
```
class DCGaussianMixtureFactor; // forward declaration
```
above?  It should return a shared pointer, so really `boost::shared_ptr<DCGaussianMixtureFactor>`  should just work.

Reply:
I actually ran it with the forward declaration and got a bunch of compiler erros, but I did a `DCGaussianMixtureFactor::shared_ptr` which is the wrong thing to do. Let me try again and resolve it with full certainty.

Reply:
That worked!! Yay